src/plan_agg_bookend.c

```diff
@@ -0,0 +1,718 @@
+static void first_last_qp_callback(PlannerInfo *root, void *extra);
+static Node *mutate_aggref_node(Node *node, MutatorContext *context);
+static void replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path);
+
+/*
+ * mutate_aggref_node
+ *
+ * Mutator function used by recursive `expression_tree_mutator`
+ * to replace Aggref node with Param node
+ */
+Node *
+mutate_aggref_node(Node *node, MutatorContext *context)
+{
+	if (node == NULL)
+		return NULL;
+	if (IsA(node, Aggref))
+	{
+		Aggref	   *aggref = (Aggref *) node;
+
+		/* See if the Aggref should be replaced by a Param */
+		if (context->mm_path != NULL &&
+			list_length(aggref->args) == 2)
+		{
+			TargetEntry *curTarget = (TargetEntry *) linitial(aggref->args);
+			ListCell   *cell;
+
+			foreach(cell, context->mm_path->mmaggregates)
+			{
+				MinMaxAggInfo *mminfo = (MinMaxAggInfo *) lfirst(cell);
+
+				if (mminfo->aggfnoid == aggref->aggfnoid &&
+					equal(mminfo->target, curTarget->expr))
+					return (Node *) copyObject(mminfo->param);
+			}
+		}
+	}
+	return expression_tree_mutator(node, mutate_aggref_node, (void *) context);
+}
+
+/*
+ * replace_aggref_in_tlist
+ *
+ * If MinMaxAggPath is chosen, instead of running aggregate
+ * function we will execute subquery that we've generated. Since we
+ * use subquery we need to replace target list Aggref node with Param
+ * node. Param node passes output value from the subquery.
+ *
+ */
+void
+replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path)
+{
+	MutatorContext context;
+
+	context.mm_path = minmaxagg_path;
+
+	((Path *) minmaxagg_path)->pathtarget->exprs = (List *) mutate_aggref_node((Node *) ((Path *) minmaxagg_path)->pathtarget->exprs, (void *) &context);
+}
+
+/* Stores function id (FIRST/LAST) with proper comparison strategy */
+typedef struct FuncStrategy
+{
+	Oid			func_oid;
+	StrategyNumber strategy;
+} FuncStrategy;
+
+static Oid	first_last_arg_types[] = {ANYELEMENTOID, ANYOID};
+
+struct FuncStrategy first_func_strategy = {.func_oid = InvalidOid,.strategy = BTLessStrategyNumber};
+struct FuncStrategy last_func_strategy = {.func_oid = InvalidOid,.strategy = BTGreaterStrategyNumber};
+
+static Oid
+get_function_oid(char *name, int nargs, Oid arg_types[])
+{
+	FuncCandidateList func_candidates;
+
+	func_candidates = FuncnameGetCandidates(list_make1(makeString(name)), nargs, NIL, false, false, false);
+	while (func_candidates != NULL)
+	{
+		if (func_candidates->nargs == nargs && function_types_equal(func_candidates->args, arg_types, nargs))
+			return func_candidates->oid;
+		func_candidates = func_candidates->next;
+	}
+	elog(ERROR, "failed to find function %s with %d args", name, nargs);
+}
+
+static FuncStrategy *
+initialize_func_sortop(FuncStrategy *func_strategy, char *name, int nargs, Oid arg_types[])
+{
+	func_strategy->func_oid = get_function_oid(name, nargs, arg_types);
+	return func_strategy;
+}
+
+static FuncStrategy *
+get_func_strategy(Oid func_oid)
+{
+	if (first_func_strategy.func_oid == InvalidOid)
```

Comment:
nit: shouldn't we do one check for initialization and initialize both. The two if's don't seem to make sense.

Reply:
Logically I think this is more correct since one can be initialized and other not.